### ISSUE TITLE
ctrl-c to cancel an encode doesn't work

### ISSUE THREAD

**Issue by RX14** [deterenkelt/Nadeshiko#14]
To reproduce:

```
nadeshiko ~/somefile.mkv
^C
```

and it carries on regardless.

ctrl-z doesn't work either

it's super annoying

**Comment by deterenkelt** [comment]
Try to press `^C` twice?

**Comment by RX14** [comment]
Doesn't work however many times I press it

**Comment by deterenkelt** [comment]
I think that I have pinpointed the problem, even though I still can’t reproduce it. Please check if the [new version](https://github.com/deterenkelt/Nadeshiko/releases/tag/v2.3.22) solves the problem.

**Comment by deterenkelt** [comment]
Also the reaction to SIGINT is a bit tardy, especially when CPU gets under 100% load. I’ve observed that
- before the first pass, nadeshiko.sh quits after `^C` immediately;
- during the first pass – after around 1–2 seconds of delay;
- during the second pass – only after 5 seconds (complete pass would last 72 seconds).

**Comment by RX14** [comment]
I updated to 2.4.1 and it's still happening

ctrl-c fails during any part of the script.

**Comment by deterenkelt** [comment]
Since the issue is not reproducible, I can only make guesses why it doesn’t work on your side.

Guess 1: the local repo is not fully updated. Try cloning to a separate directory and running with the default configuration (move ~/.config/nadeshiko somewhere temporarily)?

Guess 2: it’s the issue with the terminal emulator. Try another?

Guess 3: `tee` is broken or not accepting the `-i` key to ignore signals. This may be a bug, an old version or specifics of `tee` from BSD userland. The `tee` on my side has version 8.29 and belongs to GNU coreutils.

**Comment by RX14** [comment]
@deterenkelt this is on an up-to-date arch install, using my nadeshiko AUR package: https://aur.archlinux.org/packages/nadeshiko/. I can reproduce it on a fresh arch install. I also tried another terminal emulator but ctrl-c is pretty universal.

My tee is:

```
$ tee --version
tee (GNU coreutils) 8.30
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by Mike Parker, Richard M. Stallman, and David MacKenzie.
```

**Comment by deterenkelt** [comment]
This is indeed weird, but I have no idea what else might go wrong. I’m sorry.

The culprit, as I believe, is [this line](https://github.com/deterenkelt/Nadeshiko/blob/c23fd54c1547f3af3923c42ed2f7ae9712828279/lib/bahelite/bahelite_logging.sh#L74). While sending `^C` to find the problem, I found, that nadeshiko.sh just silently dies instead of catching an error and processing it – as it should, because everything possible is trapped. The reason for this silent failing without an error was because the nadeshiko’s self process was not in foreground enough (ehehe…) to receive signals, so they killed `tee` instead, and the nadeshiko process received therefore SIGPIPE, that I got in console when echoing `$?`. After I made `tee` pass signals, nadeshiko started to receive and trap them properly. Errors were shown, the exit code was 5 and not 141 (=128+13, the 13 being SIGPIPE).

See, if commenting the exec call could be a workaround for the time being. This is the worst, though, as it would basically disable any logging after the launch.

---

If the source is really untouched and nobody’s trying to make this code work under ksh or something, then the general problem is within pipes, the subshell or the substitute file descriptors for `>( … )` I/O. I know that when pipes are involved, there is a difference which command would run in the mother shell, and which – in a subshell. It may be the first or the last in the piping chain depending on whether “lastpipe” option is set. Check it with `shopt -p lastpipe`? It is disabled on my side. Though I doubt that it may affect the behaviour of `>( … )`, but who knows? There may be analogous option. More uncertain seems some OS setting in /etc/sysctl.

**Comment by RX14** [comment]
I made an archlinux VM for you to test this with yourself: `ssh -p 40076 deterenkelt@rx14.co.uk`. Nadeshiko is already installed and you have sudo.

**Comment by deterenkelt** [comment]
Ok, thank you, I’ll return to this issue, once I deal with #15 and #16.

**Comment by deterenkelt** [comment]
I’ve tried to dig into it today, the summary of what I discovered is below.

1. I’ve observed the lack of presence of any nadeshiko processes after pressing `^C` multiple times, but seen ffmpeg running and hogging CPU. This makes me think, that the shell running tee receives SIGINT and dies, the shell running nadeshiko receives SIGPIPE or HUP and then dies too, while ffmpeg continues to run, but `^C` sent to terminal seem to be directed to an already dead shell, hence they’re like going into a black hole. I can’t explain how this may happen.

2. The are no differences in set and shopt option sets between my host and that vm, that would affect subshells, pipes or anything like that. Shopt sets are identical.

3. This may be a bash-5 specific bug, can’t update to it right now to verify. But there’s a highly suspicious note (m) in [changelog](https://tiswww.case.edu/php/chet/bash/NEWS).

4. Setting `shopt -s compat44` doesn’t seem to have any effect, but according to the manual, it only affects the behaviour related to BASH_ARGC and ARGV variables.

5. I’ve discovered another semi-critical bug, but it is not related to the issue.

Was the issue first observed with the bash-5 update? Would be interesting to downgrade it back to 4.4.

**Comment by RX14** [comment]
@deterenkelt looking at htop, pressing ctrl-C definitely leaves nadeshiko running

**Comment by deterenkelt** [comment]
Fixed in [v2.5.11](/deterenkelt/Nadeshiko/releases/tag/v2.5.11).

**Comment by RX14** [comment]
Thanks!
